src/lib/Normalize.hs

```diff
@@ -42,19 +44,48 @@ normalize expr = case expr of
     env <- normalizeDecl decl
     extendR env $ normalize body
   FVar v -> lookupVar v
+  FPrimExpr (OpExpr (SumCase sumVal lBody rBody)) -> do
+    ~(Tup [side, lInput, rInput]) <- unpackRec =<< normalize sumVal
+    lOutput <- app lBody lInput
+    rOutput <- app rBody rInput
+    emit $ Select (getType lOutput) side lOutput rOutput
+    where
+      app body arg =
+        let lam = FPrimExpr $ ConExpr $ Lam NonLin noEffect body in
+        normalize $ FPrimExpr $ OpExpr $ App NonLin lam $ toAtomicFExpr arg
   FPrimExpr (OpExpr op) ->
-    traverseExpr op substNorm normalize normalizeLam >>= emit
+    traverseExpr op substType normalize normalizeLam >>= emit
+  FPrimExpr (ConExpr (SumCon ty' e' side)) -> do
+    other <- arbitraryValue <$> substType ty'
+    value <- normalize e'
+    return $ Con $ RecCon $ Tup $ case side of
+      Left  _ -> [Con $ Lit $ BoolLit True, value, other]
+      Right _ -> [Con $ Lit $ BoolLit False, other, value]
   FPrimExpr (ConExpr con) ->
-    liftM Con $ traverseExpr con substNorm normalize normalizeLam
+    liftM Con $ traverseExpr con substType normalize normalizeLam
   Annot    e _ -> normalize e
   SrcAnnot e _ -> normalize e
 
+arbitraryValue :: Type -> Atom
+arbitraryValue (TabTy a b)                = Con $ AFor a $ arbitraryValue b
+arbitraryValue (TC (BaseType RealType))   = Con $ Lit $ RealLit 1.0
+arbitraryValue (TC (BaseType IntType))    = Con $ Lit $ IntLit 1
+arbitraryValue (TC (BaseType BoolType))   = Con $ Lit $ BoolLit False
+arbitraryValue (TC (BaseType StrType))    = Con $ Lit $ StrLit ""
+arbitraryValue (TC (RecType r))           = Con $ RecCon $ fmap arbitraryValue r
+-- XXX: This might not be strictly legal, because those types might have no inhabitants!
+arbitraryValue t@(TC (IndexRange _ _ _))  = Con $ AsIdx t $ Con $ Lit $ IntLit 0
+arbitraryValue t@(TC (IntRange _ _))      = Con $ AsIdx t $ Con $ Lit $ IntLit 0
+arbitraryValue t                          = Con $ AsIdx t $ Con $ Lit $ IntLit 0
```

Comment:
This is a hack, because inside `restructure` we have constructors like `Left @n (i, j) : Either (n, m) n`, and the only thing we know about `n` is that it's an index set. In this case it is valid to generate an expression like this one (of course under the assumption that the index set will end up being non-empty, but that's also a problem in the previous PR...), instead of failing with an error. I'm not sure what's the best way to get my hands on the class environment at this point. Any hints?